tf_quant_finance/models/heston/approximations/asian_prices.py

```diff
@@ -0,0 +1,509 @@
+        sampling_times, dtype=dtype, name='sampling_times')
+    sampling_times = tf.broadcast_to(
+        sampling_times, [sampling_times.shape[0], strikes.shape[0]])
+
+    # Fail assertion if sampling comes after expiry
+    if averaging_frequency == AveragingFrequency.DISCRETE:
+      tf.debugging.assert_equal(
+        tf.math.maximum(sampling_times[-1], expiries), expiries,
+        'Sampling times cannot occur after expiry times'
+      )
+
+    variances = tf.convert_to_tensor(variances, dtype=dtype, name='variances')
+    variances = tf.broadcast_to(variances, strikes.shape)
+    mean_reversion = tf.convert_to_tensor(
+        mean_reversion, dtype=dtype, name='mean_reversion')
+    mean_reversion = tf.broadcast_to(mean_reversion, strikes.shape)
+    theta = tf.convert_to_tensor(theta, dtype=dtype, name='theta')
+    theta = tf.broadcast_to(theta, strikes.shape)
+    volvol = tf.convert_to_tensor(volvol, dtype=dtype, name='volvol')
+    volvol = tf.broadcast_to(volvol, strikes.shape)
+    rho = tf.convert_to_tensor(rho, dtype=dtype, name='rho')
+    rho = tf.broadcast_to(rho, strikes.shape)
+
+    if discount_rates is not None:
+      discount_rates = tf.convert_to_tensor(
+          discount_rates, dtype=dtype, name='discount_rates')
+      discount_factors = tf.exp(-discount_rates * expiries)
+    elif discount_factors is not None:
+      discount_factors = tf.convert_to_tensor(
+          discount_factors, dtype=dtype, name='discount_factors')
+      discount_rates = -tf.math.log(discount_factors) / expiries
+    else:
+      discount_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='discount_rates')
+      discount_factors = tf.convert_to_tensor(
+          1.0, dtype=dtype, name='discount_factors')
+
+    if dividend_rates is None:
+      dividend_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='dividend_rates')
+
+    if forwards is not None:
+      forwards = tf.convert_to_tensor(forwards, dtype=dtype, name='forwards')
+      spots = forwards * tf.exp(-(discount_rates - dividend_rates) * expiries)
+    else:
+      spots = tf.convert_to_tensor(spots, dtype=dtype, name='spots')
+      forwards = spots * tf.exp((discount_rates - dividend_rates) * expiries)
+
+    forwards = tf.broadcast_to(forwards, strikes.shape)
+    spots = tf.broadcast_to(spots, strikes.shape)
+
+    #  To account for seasoning, we keep pricing time fixed at t=0 and adjust
+    #  the pricing parameters as follows:
+    #    - calculate prefactor running_prod ^ (1/#_total_fixings)
+    #    - strikes \to strikes / prefactor
+    #    - k_star \to #_past_fixings
+    #    - prepend #_past_fixings negative values to sampling times
+    #    - prices \to prefactor * prices
+    #
+    #  This is slightly different to the approach in the paper, where instead
+    #  t is allowed to increase above 0 as pricing time moves forwards
+
+    if past_fixings is None:
+      running_accumulator = tf.convert_to_tensor(1.0, dtype=dtype)
```

Comment:
nit: here use tf.constant as 1.0 is a constant

Reply:
Done